quint/src/builtin.qnt

```diff
@@ -852,6 +852,11 @@ module builtin {
   /// ```
   action repeated(a, n): (bool, int) => bool
 
+  /// `a.fail()` evaluates to `true` if and only if action `a` evaluates to `false`.
```

Comment:
I’m a bit surprised fail has to be a built in. Why couldn’t it be defined as a negation?

Reply:
Great question. I think in TLA+ it could be defined as `~ENABLED(A)`, that is the condition that `A` is disabled. The issue is that it is not simply a negation, unless you drop all assignments, which is quite hard to do in TLA+ and relatively easy in Quint. Moreover, the more I am looking and actions and runs, the more I believe we need its own type system for action results. It should be either binary like `Executed` and `NotExecuted`. Or better ternary, like `Executed`, `NotExecuted`, `Fail`. There is a tricky distinction between a disabled action and the one that failed, because it has run into some inconsistency inside the action (the typical example is an `assert`).

I have not thought about this much, but it seems that a ternary system would help us to distinguish between preconditions and failures.